### DISCUSSION TITLE
Any ideas to filter by multiple relation

### DISCUSSION THREAD

**Discussion by bigint** [prisma/prisma#9287] [Q&A] [answered]
I want to return the following user's posts and also the topic I follow, the below query returns only the following user's query, not the following topic.

Any ideas on how to handle it?

```js
return await db.post.findMany({
    where: {
        AND: {
            user: {
                id: {
                    in: [
                        ...userFollowing?.following.map((user: any) => user.id),
                        session?.userId
                    ]
                },
                spammy: false
            },
            topics: {
                every: {
                    topicId: {
                        in: [...userTopics?.topics.map((topic: any) => topic.id)]
                    }
                }
            }
        },
        hidden: false
    },
})
```

**Comment by ryands17** [comment, marked as answer]
Using `some` instead of every in `topics` should make this work.

**Reply by bigint** [reply]
@ryands17 using `some` returns only one post 🤔 

![image](https://user-images.githubusercontent.com/69431456/133559172-fc2a5331-b013-472e-b1d0-909ce5a42e5c.png)

**Reply by bigint** [reply]
FYI here is the schema

<details>
<summary>
schema.prisma
</summary>
datasource db {
  provider          = "mysql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
  planetScaleMode   = true
}

generator client {
  provider        = "prisma-client-js"
  binaryTargets   = ["native", "debian-openssl-1.1.x"]
  previewFeatures = ["filterJson", "planetScaleMode"]
}

generator giraphql {
  provider = "prisma-giraphql-types"
}

model User {
  id             String  @id @default(cuid())
  username       String  @unique @default(cuid())
  email          String  @unique
  hashedPassword String

  // Flags
  spammy     Boolean @default(false)
  isVerified Boolean @default(false)
  isStaff    Boolean @default(false)
  inWaitlist Boolean @default(true)

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relations
  posts            Post[]
  products         Product[]
  communities      Community[]
  sessions         Session[]
  profile          Profile?
  invite           Invite?
  integrations     Integration?
  likes            Like[]
  badges           Badge[]
  topics           Topic[]
  following        User[]         @relation("follows")
  followedBy       User[]         @relation("follows")
  myNotification   Notification[] @relation("receiver")
  sentNotification Notification[] @relation("dispatcher")

  // Attributes
  @@index([username, email])
  @@map("users")
}

model Session {
  id        String  @id @default(cuid())
  isStaff   Boolean @default(false)
  userAgent String?

  // Timestamps
  createdAt DateTime @default(now())
  expiresAt DateTime @default(now())

  // Relations
  user   User   @relation(fields: [userId], references: [id])
  userId String

  // Attributes
  @@index([userAgent, userId])
  @@map("sessions")
}

model Profile {
  id       String  @id @default(cuid())
  name     String
  avatar   String?
  cover    String?
  bio      String?
  location String?
  website  String?
  discord  String?
  github   String?
  twitter  String?

  // Relations
  user   User   @relation(fields: [userId], references: [id])
  userId String

  // Attributes
  @@index([userId])
  @@map("profiles")
}

model Invite {
  id        String  @id @default(uuid())
  code      String?
  usedTimes Int     @default(0)

  // Timestamps
  createdAt DateTime @default(now())

  // Relations
  user   User   @relation(fields: [userId], references: [id])
  userId String

  // Attributes
  @@index([code, userId])
  @@map("invites")
}

model Post {
  id          String   @id @default(cuid())
  title       String?
  body        String
  done        Boolean  @default(false)
  type        PostType @default(POST)
  attachments Json?
  hidden      Boolean  @default(false)

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  deletedAt DateTime?

  // Relations
  user      User        @relation(fields: [userId], references: [id])
  userId    String
  product   Product?    @relation(fields: [productId], references: [id])
  productId String?
  likes     Like[]
  topics    PostTopic[]

  parentId String?
  parent   Post?   @relation("replies", fields: [parentId], references: [id])
  replies  Post[]  @relation("replies")

  // Attributes
  @@index([userId, productId, parentId])
  @@map("posts")
}

model PostTopic {
  id String @id @default(cuid())

  // Relations
  post    Post?   @relation(fields: [postId], references: [id])
  postId  String?
  topic   Topic?  @relation(fields: [topicId], references: [id])
  topicId String?

  // Attributes
  @@index([postId, topicId])
  @@map("post_topics")
}

model Topic {
  id          String  @id @default(cuid())
  name        String  @unique
  image       String?
  description String?

  // Timestamps
  createdAt DateTime @default(now())

  // Relations
  posts PostTopic[]
  users User[]

  // Attributes
  @@index([name])
  @@map("topics")
}

model Like {
  id String @id @default(cuid())

  // Timestamps
  createdAt DateTime @default(now())

  // Relations
  user         User           @relation(fields: [userId], references: [id])
  userId       String
  post         Post?          @relation(fields: [postId], references: [id])
  postId       String?
  notification Notification[] @relation("like")

  // Attributes
  @@unique([userId, postId], name: "likeIdentifier")
  @@index([postId, userId])
  @@map("likes")
}

model Product {
  id          String  @id @default(cuid())
  name        String
  slug        String  @unique
  description String?
  avatar      String?

  // Social
  website     String?
  producthunt String?
  discord     String?
  github      String?
  twitter     String?

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relations
  user   User?   @relation(fields: [userId], references: [id])
  userId String?
  posts  Post[]

  // Attributes
  @@index([slug, userId])
  @@map("products")
}

model Community {
  id          String  @id @default(cuid())
  name        String
  slug        String  @unique
  description String?
  avatar      String?

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relations
  users User[]

  // Attributes
  @@index([slug])
  @@map("communities")
}

model Badge {
  id          String  @id @default(cuid())
  name        String
  image       String
  description String?

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relations
  users User[]

  // Attributes
  @@map("badges")
}

model Notification {
  id      String           @id @default(cuid())
  message String?
  type    NotificationType

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relations
  receiver     User    @relation("receiver", fields: [receiverId], references: [id])
  receiverId   String
  dispatcher   User    @relation("dispatcher", fields: [dispatcherId], references: [id])
  dispatcherId String
  like         Like?   @relation("like", fields: [likeId], references: [id])
  likeId       String?

  // Attributes
  @@index([receiverId, dispatcherId, likeId])
  @@map("notifications")
}

model Integration {
  id                  String  @id @default(uuid())
  wakatimeAPIKey      String?
  spotifyRefreshToken String?

  // Timestamps
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Relations
  user   User   @relation(fields: [userId], references: [id])
  userId String

  // Attributes
  @@index([userId])
  @@map("integrations")
}

// Types
enum PostType {
  POST
  TASK
  QUESTION
  REPLY
}

enum NotificationType {
  POSTLIKE
  FOLLOW
}
</details>

**Reply by ryands17** [reply]
It must be a combination so `AND` would return only the ones that match.

**Reply by bigint** [reply]
Should I need to remove `AND`? Even after removing `AND` didn't work :(

```js
return await db.post.findMany({
    where: {
      type: type === 'ALL' ? undefined : (type as PostType),
      user: {
        id: {
          in: [
            ...userFollowing.following.map((user: any) => user.id),
            session?.userId
          ]
        },
        spammy: false
      },
      topics: {
        some: {
          topic: {
            id: { in: [...userTopics?.topics.map((topic: any) => topic.id)] }
          }
        }
      },
      hidden: false
    },
  })
```

**Reply by bigint** [reply]
Woah seems like this is working 😄 

```js
return await db.post.findMany({
    ...query,
    where: {
      type: type === 'ALL' ? undefined : (type as PostType),
      OR: [
        {
          user: {
            id: {
              in: [
                ...userFollowing.following.map((user: any) => user.id),
                session?.userId
              ]
            },
            spammy: false
          }
        },
        {
          topics: {
            some: {
              topic: {
                id: {
                  in: [...userTopics?.topics.map((topic: any) => topic.id)]
                }
              }
            }
          }
        }
      ],

      hidden: false
    },
    orderBy: { createdAt: 'desc' }
  })
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
